src/lightning/pytorch/trainer/trainer.py

```diff
@@ -464,6 +472,10 @@ def fit(
                 If resuming from mid-epoch checkpoint, training will start from the beginning of the next epoch.
 
             datamodule: An instance of :class:`~lightning.pytorch.core.datamodule.LightningDataModule`.
+
+        Raises:
+            TypeError:
+                If ``model`` is not `LightningModule` or `torch._dynamo.OptimizedModule`.
```

Comment:
How are the exceptions included decided? This method doesn't explicitly show the `TypeError`, it's only inside `_maybe_unwrap_optimized`, and it's gated by some conditions. Would you mention those conditions here? How do we make sure the docstring does not become outdated?

These limitations make me prefer that we don't add these "Raises" blocks personally.

Reply:
I have added those conditions as well here.

Reply:
regarding what exceptions to be included, yes, this was a concern I also had while writing this. 

as you said, if we add all exceptions from internal methods too, then there is high chance that the docstring can get outdated, when someone modifies the internal method and forgets to modify the docstring of all public/high level methods calling it.

On the other hand, if I just add the exceptions in the high level public function, then it's not really telling the whole story right. Also if in future those exceptions are refracted into some internal methods, the docstring needs to be removed or we will have this inconsistency. 

One cleaner solution I can think of is to put references in the main method docstring referring the docstring of internal methods. But even that can get little messy with too many references.

Reply:
Exactly, there's no good solution. So I would suggest not doing this. But if other reviewers are fine with these limitations (cc @akihironitta @Borda @awaelchli) then feel free to resolve

Reply:
I also recommend not doing this, it adds a lot of text but little value.